src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala

```diff
@@ -1,15 +1,38 @@
 package org.broadinstitute.dsde.workbench.sam.api
 
 import akka.http.scaladsl.server._
-import akka.http.scaladsl.server.Directives.headerValueByName
-import org.broadinstitute.dsde.workbench.model.{WorkbenchUserEmail, WorkbenchUserId}
+import akka.http.scaladsl.server.Directives.{headerValueByName, onSuccess}
+import org.broadinstitute.dsde.workbench.model._
 import org.broadinstitute.dsde.workbench.sam.model.UserInfo
 
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+
+
+
 trait StandardUserInfoDirectives extends UserInfoDirectives {
+  val petSAdomain = "\\S+@\\S+.iam.gserviceaccount.com"
+
   def requireUserInfo: Directive1[UserInfo] = (
     headerValueByName("OIDC_access_token") &
     headerValueByName("OIDC_CLAIM_user_id") &
     headerValueByName("OIDC_CLAIM_expires_in") &
     headerValueByName("OIDC_CLAIM_email")
-  ).tmap { case (token, userId, expiresIn, email) => UserInfo(token, WorkbenchUserId(userId), WorkbenchUserEmail(email), expiresIn.toLong) }
+  ) tflatMap {
+    case (token, userId, expiresIn, email) => {
+      onSuccess(getWorkbenchUserEmailId(email)).map {
+        case Some(resourceType) => UserInfo(token, resourceType.id, resourceType.email, expiresIn.toLong)
+        case None => UserInfo(token, WorkbenchUserId(userId), WorkbenchUserEmail(email), expiresIn.toLong)
+      }
+    }
+  }
+
+  private def isPetSA(email:String) = email.matches(petSAdomain)
```

Comment:
This works, but it causes the pattern to be recompiled each time, which could have a performance impact. You can compile the pattern once like so:

```
val petSAdomain = "\\S+@\\S+\\.iam\\.gserviceaccount\\.com".r

private def isPetSA(email: String) = {
  petSAdomain.pattern.matcher(email).matches
}
```